agents/agents-ext/src/commonMain/kotlin/ai/koog/agents/ext/tool/search/RegexSearchTool.kt

```diff
@@ -0,0 +1,154 @@
+import ai.koog.rag.base.files.toPosition
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.emitAll
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.mapNotNull
+import kotlinx.coroutines.flow.take
+import kotlinx.coroutines.flow.toList
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+/**
+ * Regular expression based content search tool.
+ *
+ * Use to find occurrences of a regex pattern across text files under a path.
+ */
+public class RegexSearchTool<Path>(
+    private val fs: FileSystemProvider.ReadOnly<Path>,
+) : Tool<RegexSearchTool.Args, RegexSearchTool.Result>() {
+
+    override val name: String = "__search_contents_by_regex__"
+    override val description: String = text {
+        +"Executes a regular expression search on folder or file contents within the specified path."
+        +"The tool returns structured results with file paths, line numbers, positions, and excerpts where the text was found."
+        +"The tool will solely return search results and does not modify any files."
+    }
+
+    @Serializable
+    public data class Args(
+        @param:LLMDescription("Absolute starting directory or file path.")
+        val path: String,
+        @param:LLMDescription("Regular expression pattern.")
+        val regex: String,
+        @param:LLMDescription("Maximum number of matching files to return (pagination).")
+        val limit: Int = 25,
+        @param:LLMDescription("Number of matching files to skip (pagination).")
+        val skip: Int = 0,
+        @SerialName("case_sensitive")
+        @param:LLMDescription("If false, performs case-insensitive matching.")
+        val caseSensitive: Boolean = false,
+    )
+
+    @Serializable
+    public data class Result(val entries: List<FileSystemEntry.File>, val original: String)
+
+    override val argsSerializer: KSerializer<Args> = Args.serializer()
+    override val resultSerializer: KSerializer<Result> = Result.serializer()
+
+    override suspend fun execute(args: Args): Result {
+        val path = fs.fromAbsolutePathString(args.path)
+        val matches = search(path, args.regex, args.limit, args.skip, args.caseSensitive).toList()
+        return Result(matches, original = args.regex)
+    }
+
+    private suspend fun search(
+        path: Path,
+        pattern: String,
+        limit: Int,
+        skip: Int,
+        caseSensitive: Boolean,
+        linesAroundSnippet: Int = 2,
+    ): Flow<FileSystemEntry.File> {
+        val options = mutableSetOf<RegexOption>()
+        if (!caseSensitive) options.add(RegexOption.IGNORE_CASE)
+
+        return searchByRegex(
+            fs = fs,
+            start = path,
+            regex = Regex(pattern, options)
+        )
+            .drop(skip)
+            .take(limit)
+            .mapNotNull { match ->
+                val content = fs.readText(match.file)
```

Comment:
You're reading each file twice - once to filter the lines in `searchByRegex`, and once here. This is not optimal and can lead to perfomance problems on a large set or for large files. Consider rewriting the logic to read and save relevant chunks to memory only once